source/common/config/new_grpc_mux_impl.cc

```diff
@@ -0,0 +1,238 @@
+                                        const std::set<std::string>& resources,
+                                        SubscriptionCallbacks& callbacks,
+                                        std::chrono::milliseconds init_fetch_timeout) {
+  if (watch == nullptr) {
+    return addWatch(type_url, resources, callbacks, init_fetch_timeout);
+  } else {
+    updateWatch(type_url, watch, resources);
+    return watch;
+  }
+}
+
+void NewGrpcMuxImpl::removeWatch(const std::string& type_url, Watch* watch) {
+  updateWatch(type_url, watch, {});
+  auto entry = subscriptions_.find(type_url);
+  if (entry == subscriptions_.end()) {
+    ENVOY_LOG(error, "removeWatch() called for non-existent subscription {}.", type_url);
+    return;
+  }
+  entry->second->watch_map_.removeWatch(watch);
+}
+
+void NewGrpcMuxImpl::pause(const std::string& type_url) { pausable_ack_queue_.pause(type_url); }
+
+void NewGrpcMuxImpl::resume(const std::string& type_url) {
+  pausable_ack_queue_.resume(type_url);
+  trySendDiscoveryRequests();
+}
+
+bool NewGrpcMuxImpl::paused(const std::string& type_url) const {
+  return pausable_ack_queue_.paused(type_url);
+}
+
+void NewGrpcMuxImpl::onDiscoveryResponse(
+    std::unique_ptr<envoy::api::v2::DeltaDiscoveryResponse>&& message) {
+  ENVOY_LOG(debug, "Received DeltaDiscoveryResponse for {} at version {}", message->type_url(),
+            message->system_version_info());
+  auto sub = subscriptions_.find(message->type_url());
+  if (sub == subscriptions_.end()) {
+    ENVOY_LOG(warn,
+              "Dropping received DeltaDiscoveryResponse (with version {}) for non-existent "
+              "subscription {}.",
+              message->system_version_info(), message->type_url());
+    return;
+  }
+  kickOffAck(sub->second->sub_state_.handleResponse(*message));
+}
+
+void NewGrpcMuxImpl::onStreamEstablished() {
+  for (auto& sub : subscriptions_) {
+    sub.second->sub_state_.markStreamFresh();
+  }
+  trySendDiscoveryRequests();
+}
+
+void NewGrpcMuxImpl::onEstablishmentFailure() {
+  // If this happens while Envoy is still initializing, the onConfigUpdateFailed() we ultimately
+  // call on CDS will cause LDS to start up, which adds to subscriptions_ here. So, to avoid a
+  // crash, the iteration needs to dance around a little: collect pointers to all
+  // SubscriptionStates, call on all those pointers we haven't yet called on, repeat if there are
+  // now more SubscriptionStates.
+  absl::flat_hash_map<std::string, DeltaSubscriptionState*> all_subscribed;
+  absl::flat_hash_map<std::string, DeltaSubscriptionState*> already_called;
+  do {
+    for (auto& sub : subscriptions_) {
+      all_subscribed[sub.first] = &sub.second->sub_state_;
+    }
+    for (auto& sub : all_subscribed) {
+      if (already_called.insert(sub).second) { // insert succeeded ==> not already called
+        sub.second->handleEstablishmentFailure();
+      }
+    }
+  } while (all_subscribed.size() != subscriptions_.size());
```

Comment:
In the first loop: In what scenario that this condition is false? handleEstablishmentFailure() may mutate `subscriptions_`?

Reply:
Yes, described by the comment above.

Reply:
By going through handleEstablishmentFailure overrides I understand the case that CDS as well as other xDS could respond ready() as reaction to failure.

Suppose the newly added subscription is a LDS subscription. My follow up question is why it is critical to invoke `handleEstablishmentFailure()`? You mentioned it as `crash`.
To avoid that crash you check size equal since your assumption is by end of handleEstablishmentFailure() return the new LDS subscription must put itself in `subscriptions_`, 
What if the LDS subscription comes a little later than `handleEstablishmentFailure()`? Crash?

Reply:
I'm afraid I don't entirely comprehend what you're saying, but let me try to answer what I think you're asking. This mux thing is responsible for passing on news of the failure to establish the channel along to the various xDS objects it serves, including ones that pop up as a result of telling others the news. If the LDS popped up later, it wouldn't matter, because the failure would have already happened before it popped up. (If you are thinking in terms of concurrency, all of this management code is completely single threaded (thankfully!))

Reply:
There is an recent istio change that expecting LDS respond with either all the listeners when envoy starts. 
istio assumes either 
 - all the listeners is in the first SUCCESSFUL response or,
 - in case any error, envoy give up the LDS response.
I am trying to understand if this hold true with this PR.

Questions 1 by 1:
- Without the while, why would envoy crash, according to your comment, when another listener subscription happens after `onEstablishmentFailure`?
- Is the missing notification critical? Since there is retry timer to reconnect and envoy would eventually establish successfully or invoke another `onEstablishmentFailure`

Reply:
This part of the code you're looking at is in the context of the xDS gRPC stream failing to even connect. So, there is not any LDS response at all.

Reply:
> This part of the code you're looking at is in the context of the xDS gRPC stream failing to even connect. So, there is not any LDS response at all.
Good point. There won't be any successful LDS response, so no partial response either. 

> Is the missing notification critical? 
Other people is more aggressive #8152 Establishment failure is hidden to upper layer. So `all_subscribed.size() != subscriptions_.size()` should not even happen.
I am ok if we leave it here for now and delete in the future once we are fully confident